proxy/coco/src/request/waiting_room.rs

```diff
@@ -149,16 +158,16 @@ impl<T, D> WaitingRoom<T, D> {
     /// This will return the request for the given `urn` if one exists in the `WaitingRoom`.
     ///
     /// If there is no such `urn` then it create a fresh `Request` using the `urn` and `timestamp`
-    /// and it will return `None`.
+    /// and it will return the new `Request`.
     pub fn request(&mut self, urn: RadUrn, timestamp: T) -> Option<SomeRequest<T>>
     where
         T: Clone,
     {
         match self.get(&urn) {
             None => {
                 let request = SomeRequest::Created(Request::new(urn.clone(), timestamp));
-                self.requests.insert(urn.id, request);
-                None
+                self.requests.insert(urn.id, request.clone());
+                Some(request)
```

Comment:
Ok, so here I'd like to point out some things I'd like you to think about with regards to types and semantics.
1. If we _always_ return a request then there is no need for an `Option` because we always return `Some` at the end. Essentially the function signature becomes a lie. It says that it's a possibility that this function will fail but it never does.
2. The semantics of this function was that we want to have some signal of when a request was new vs when it already existed. You've now removed those semantics and we never know this information.

I think it's good practice to keep these kinds of things in mind: "What is the type signature saying and is it true?" and "What are the semantics of this function? Am I changing them?" Sometimes you do need to change the signature and semantics but in this case, I think we only need to change the signature to use `Either`, as mentioned in a previous comment.